Memory cache modules size

maurolacy opened the pull request:
A small PR with the tools for memory profiling of compiled modules.

Not sure if you'd want to merge this, as it pollutes the `cosmwasm_vm` package with tools and deps, but publishing nevertheless for reference.

Same technique can be useful for mem / heap profiling of other stuff.

webmaster128 reviewed (commented):
This is great stuff. I'm happy to merge this as long as we can avoid unnecessary deps for contract developers.

I tried to use this on the Mac, but at least with Homebrew, [valgrind requires Linux](https://formulae.brew.sh/formula/valgrind).

I think `compile_and_use` would be better to check, as it includes the memory limiting functionality (not big but closer to the real use case). Also it would be great if we could compare

1. a Module from `compile_and_use` with
2. a Module deswerialized with a `make_runtime_store`

because 1. contains a compiler in its Store and 2. does not contain a compiler.

maurolacy commented:
> This is great stuff. I'm happy to merge this as long as we can avoid unnecessary deps for contract developers.

OK. I'll move this around, so that it can be a dev-dependency.

> 
> I tried to use this on the Mac, but at least with Homebrew, [valgrind requires Linux](https://formulae.brew.sh/formula/valgrind).
> 

[Valgrind home page](https://valgrind.org/) mentions Mac OS X (10.12) is supported. It may be a litle tricky to compile and get it running, though.

> I think `compile_and_use` would be better to check, as it includes the memory limiting functionality (not big but closer to the real use case). Also it would be great if we could compare
> 
>     1. a Module from `compile_and_use` with
> 
>     2. a Module deswerialized with a `make_runtime_store`
> 

Yeah, I was thinking about extending the tool with a Module from deserialization. Will work on it.

> 
> because 1. contains a compiler in its Store and 2. does not contain a compiler.

Good. Will check that.

maurolacy commented on packages/vm/README.md:
````diff
@@ -71,6 +71,20 @@ cd packages/vm
+
+```
+cd packages/vm
+cargo build
````
I tried it, and can confirm that the (global) memory footprint / layout stays the same.

It's tricky to do detailed memory profiling of release builds, because without symbols it's more difficult to identify the relevant parts. In `module_size.sh` we are relying on the function name / source file line number to do the aggregation, and that info is lost for release builds.

I can try to do a detailed analysis of a release build, but so far everything indicates the memory footprint will stay the same.

webmaster128 replied:
Good to know, thanks.

You can enable debug symbols for release builds as well: https://stackoverflow.com/questions/38803760/how-to-get-a-release-build-with-debugging-information-when-using-cargo. The second answer looks good because it does not change the project's default release profile.

maurolacy replied:
Nice. Will give it a shot.

maurolacy replied:
This works great, and the resulting binary runs much faster. It produces exactly the same results for memory footprint.

maurolacy commented:
> [Valgrind home page](https://valgrind.org/) mentions Mac OS X (10.12) is supported. It may be a litle tricky to compile and get it running, though.

It seems latest / recent versions work: [https://stackoverflow.com/questions/40650338/valgrind-on-macos-sierra](https://stackoverflow.com/questions/40650338/valgrind-on-macos-sierra).

webmaster128 commented:
> [Valgrind home page](https://valgrind.org/) mentions Mac OS X (10.12) is supported. It may be a litle tricky to compile and get it running, though.

No worries, I have a Linux machine available for running this. No need to invest time into Mac.

maurolacy commented:
> > [Valgrind home page](https://valgrind.org/) mentions Mac OS X (10.12) is supported. It may be a litle tricky to compile and get it running, though.
> 
> No worries, I have a Linux machine available for running this. No need to invest time into Mac.

OK. I don't have a Mac around so I cannot test this directly anyway.

webmaster128 commented on packages/vm/src/bin/module_size.rs:
```diff
@@ -0,0 +1,46 @@
+use std::fs::File;
```
I guess this file was supposed to be `bin/module_size.rs` not `src/bin/module_size.rs` right? The binaries (like benchs, examples, tests) are not in `src`. Anyways, this will change when turned into an example.

maurolacy replied:
No, this is a just way to compile an `bin` instead of a `lib`. Not sure what happens if I put it in `./bin` directly... wil try.

There are other ways of having mixed or multiple build artifacts with a single `Cargo.toml`, (i. e. having both `[lib]` and `[[bin]]` sections.
I like this one because it doesn't require changes to the `Cargo.toml`.

webmaster128 replied:
> There are other ways of having mixed or multiple build artifacts with a single `Cargo.toml`, (i. e. having both `[lib]` and `[[bin]]` sections.

Yeah, this is the way I was referring to. It clearly separates the library from binaries build around it.

webmaster128 reviewed (commented):
Very nice. I just tested the latest version and get a small value for the unserialized module (< 800KB):

```
$ ./examples/module_size.sh ./testdata/hackatom.wasm
==10935== Massif, a heap profiler
==10935== Copyright (C) 2003-2017, and GNU GPL'd, by Nicholas Nethercote
==10935== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==10935== Command: ./examples/../../../target/debug/examples/module_size ./testdata/hackatom.wasm
==10935== 
wasm size: 186096 bytes
module size (serialized): 3226369 bytes
(serialized) module size ratio: 17.34
==10935== 
module size (unserialized): 789752 bytes
``` 

Is this expected? No 3-4 megabytes anymore?

maurolacy commented:
> Is this expected? No 3-4 megabytes anymore?

Should be the same as before. Works for me here. Weird... try recompiling vm? valgrind version?

webmaster128 commented:
All recompiled. Rust 1.49.0 (same for 1.48.0). valgrind-3.15.0.

The location module_size.rs:36 is only mentioned once in a 7.99% block. But I don't know how to read those outputs. Here is the full summary:
[ms_print.log](https://github.com/CosmWasm/cosmwasm/files/5831775/ms_print.log)

maurolacy commented:
This is probably a problem with valgrind. I have valgrind-3.14.0, btw.

Try increasing the number of snapshots, to see if it makes a difference (line 7 in `module_size.sh`).

Did this worked for you before the "move module_size to example" changes?

webmaster128 commented:
> Did this worked for you before the "move module_size to example" changes?

Nope, I got roughly the same value in at all stages. Will check again on a different machine.

maurolacy commented:
> > Did this worked for you before the "move module_size to example" changes?
> 
> Nope, I got roughly the same value in at all stages. Will check again on a different machine.

Thanks for this. Seems valgrind misses entries / stack frames in some cases. In any case, it's fixed in the latest version. :-)

maurolacy commented:
> > ```
> > 1. a Module from `compile_and_use` with
> > 
> > 2. a Module deswerialized with a `make_runtime_store`
> > ```
> 
> 
> Yeah, I was thinking about extending the tool with a Module from deserialization. Will work on it.
> 
> > because 1. contains a compiler in its Store and 2. does not contain a compiler.
> 
> 
> Good. Will check that.

I've implemented this, but maybe not in the right way, as I'm using the previously compiled module. @webmaster128 how do I deserialize without compiling first? That is, without having a compiler attached to the Module?

Also, take into account this is an empirical method. Valgrind sometimes fails to report entries, and sometimes, reports them multiple times. So, this is a best-effort attempt at memory profiling / measurement. I've added some heuristics to make this more or less reliable, but still.

webmaster128 commented:
> @webmaster128 how do I deserialize without compiling first? That is, without having a compiler attached to the Module?

The compiler is part of the Engine, which is part of the Store. You cannot change the Store of a Module due to a missing API. However, if you serialize a Module, the Store is completely stripped off. So in order to get the size of a Module without compiler take a serialization and create a Module with a headless store.

webmaster128 commented on packages/vm/examples/module_size.rs:
```diff
@@ -0,0 +1,62 @@
+    mem::drop(module);
+
+    // Deserialize using make_runtime_store()
+    let store = make_runtime_store(memory_limit);
+
+    let module = unsafe { Module::deserialize(&store, &serialized) }.unwrap();
```
What about wrapping those two in a named function which we later can search? Then we get the allocations for
1. Creating the headless store
2. Creating the Module using this store
as one report. This is what we need at high level: memory consumtion of Module including store.

maurolacy replied:
Sure, but, you mean, in the vm API, or in `module_size`?

webmaster128 replied:
I mean here in this example. Just pull out a function that grups the two. If Rust does some inlining we don't want because it is the same file, we can probably use `#[inline(never)]` or something like that.

maurolacy replied:
`#[inline(never)]` was in fact needed. Thanks for that.

maurolacy commented:
> So in order to get the size of a Module without compiler take a serialization and create a Module with a headless store.

So, the code is correct as it stands now.

Curiously, the resulting module size, after deserializing from this "headless store" module, is a little bit (~0.1%) more.

maurolacy commented:
> > So in order to get the size of a Module without compiler take a serialization and create a Module with a headless store.
> 
> So, the code is correct as it stands now.
> 
> Curiously, the resulting module, after deserializing from this "headless store" module, is a little bit (~0.1%) larger.

What I'm thinking now is that we're measuring total heap allocation from these functions, but that some of those allocations could not (necessarily) be attached to the resulting `Module`.

Nevertheless, the fact that the resulting module size closely matches the serialized module size, makes me trust these results.

ethanfrey commented:
Just following along here. It would be nice to see a summary of some of the results committed (as md or txt file).

Also, it would be nice to see the results from a few different sample wasm modules (to ensure they all produce similar results and we don't hit skew based on one particular module).

You both seem to have the profiling well under control. I am just curious of the various results.

webmaster128 commented:
> Also, it would be nice to see the results from a few different sample wasm modules (to ensure they all produce similar results and we don't hit skew based on one particular module).

Once we are happy to merge this, I test it with the drand contract, which is probably the largest we built so far (700KB). But I need to update the contract first.

maurolacy commented:
I've tested it with various contracts, both optimized and not optimized.

For optimized ones, the ratio of wasm size to module size is ~18. For non-optimized ones, it's ~4. This of course is beause non-optimized contracts are much bigger. The compiled module size is the same in both cases.

ethanfrey commented:
> The compiled module size is the same in both cases.

That is interesting. So it strips out dead code somehow here as well?

maurolacy commented:
> > The compiled module size is the same in both cases.
> 
> That is interesting. So it strips out dead code somehow here as well?

You are right, non-optimized Modules are about 50% larger. I should have said "roughly" the same size.

webmaster128 commented:
Okay compiler vs. headless makes no big difference. Good to know.

The large contract (704890 bytes) is 9492793 bytes serialized (13.47x) and 9541456 bytes in memory (13.54x). So the factor is larger for small contract I guess. @maurolacy how large is the smallest contract you tested?

maurolacy commented:
> how large is the smallest contract you tested?

That would be the optimized `cw20_base` contract:
```
$ ./examples/module_size.sh ~/tmp/artifacts/cw20_base.wasm 
    Finished release [optimized] target(s) in 0.08s
==17286== Massif, a heap profiler
==17286== Copyright (C) 2003-2017, and GNU GPL'd, by Nicholas Nethercote
==17286== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==17286== Command: ./examples/../../../target/release/examples/module_size /home/mauro/tmp/artifacts/cw20_base.wasm
==17286== 
wasm size: 15721 bytes
module size (serialized): 278221 bytes
(serialized) module size ratio: 17.70
==17286== 
module size (compile_and_use): 276264 bytes
module size (module_deserialize): 270661 bytes
$ 
```

webmaster128 commented on packages/vm/examples/module_size.rs:
```diff
@@ -0,0 +1,65 @@
+
+    // Compile module
+    let memory_limit = Some(Size::mebi(10));
+    let module = compile_and_use(&wasm, memory_limit).unwrap();
```
What about wrapping this in a function too? Then it is more consistent and we don't run into issues if `compile_and_use` is used elsewhere in the script.

webmaster128 commented on packages/vm/src/lib.rs:
```diff
@@ -44,3 +44,14 @@ pub use crate::ibc_calls::{
+    //! Please don't use any of these types directly, as
+    //! they might change frequently or be removed in the future.
+
+    pub use crate::wasm_backend::compile_and_use;
+    pub use crate::wasm_backend::make_runtime_store;
```
```suggestion
    pub use crate::wasm_backend::{compile_and_use, make_runtime_store};
```

webmaster128 reviewed (commented):
Two tiny suggestions. Otherwise good to merge.

webmaster128 reviewed (approved):
Very nice
